packages/core/src/types.ts

```diff
@@ -25,12 +25,20 @@ export type RequestPolicy =
 /** How the operation has */
 export type CacheOutcome = 'miss' | 'partial' | 'hit';
 
+/** Serialisable GraphQL primitives */
+type Primitive = PrimitiveMap | number | boolean | string | null | undefined;
+
+/** A map of primitives and primitive maps */
+interface PrimitiveMap {
+  [key: string]: Primitive | Primitive[] | PrimitiveMap;
+}
+
 /** A Graphql query, mutation, or subscription. */
 export interface GraphQLRequest {
   /** Unique identifier of the request. */
   key: number;
   query: DocumentNode;
-  variables?: object;
+  variables?: PrimitiveMap;
```

Comment:
I think this change should be made to the framework bindings, e.g. `useQuery` etc, in particular by adding it to the generic there like we do below for `OperationResult`. That's because someone may be using `variables` to add some non-serialisable metadata, like `Date`, which would then be stringified by `stringifyVariables`. It may also be used to pass variables that are only used offline and then filtered out, so I'm a little uneasy about making more assumptions here

Reply:
Yeah I was thinking of the toString issue - would [adding something like this](http://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=15&pc=38#code/PQKhAIGUFMCcEsCGAbeBnRAjZ1wHFZEAHACwEUAZcIhAW3gBd4A3aNcEYAWACgGBPIrgAKdRi1wBecOAB2AV1qY44AD7hMAe005EsteDQMEsgOYGFyZAfmyAJtABm8WdDsGA3uAabIxl6YAXOAAFACU4JIAfIb+ZuAAvgDcvLygEACC4LTE4JqO1GJMrOx67jTw9MW4OUTsnLwuDHCOiADGIkUSALK5HrwyANoA1tD8wUYmpgC6waKV4qwG81USg9PLXay9RCk8Cak8bciIaOzd-ADCJ2fg-TwyMkTy2PBt3r5xpuF3A4--sGgDHksH0ACILuBjqd2C4jHoOmC9v8DvtDm1NLIjOBELAguAVospHJoAB3cAXa4w8JJIA) to the primitives help?

I'm personally in the mindset that users should be making sure that their variables should be intentionally converted beforehand but I appreciate that enforcing that isn't a great idea.

Reply:
It's unlikely 😅  We now also support passing `File` and `Blob` variables for `@urql/exchange-multipart-fetch` and cases like these make it hard for us to have a stricter type by default 🤔